platform/tls/generic/mbedtls/src/mender-tls.c

```diff
@@ -127,77 +127,90 @@ mender_tls_init(void) {
-                                                         user_provided_key_length))) {
+        mender_log_debug("Retrieving user provided authentication keys");
+
+        if (MENDER_OK != get_user_provided_keys(&user_provided_key, &user_provided_key_length)) {
             mender_log_error("Unable to get user provided authentication key");
-            goto END;
+            return MENDER_FAIL;
         }
-        /* Retrieve or generate private and public keys */
-    } else if (MENDER_OK
-               != (ret = mender_storage_get_authentication_keys(
-                       &mender_tls_private_key, &mender_tls_private_key_length, &mender_tls_public_key, &mender_tls_public_key_length))) {
-        /* Generate authentication keys */
-        mender_log_info("Generating authentication keys...");
-        if (MENDER_OK
-            != (ret = mender_tls_get_authentication_keys(
-                    &mender_tls_private_key, &mender_tls_private_key_length, &mender_tls_public_key, &mender_tls_public_key_length, NULL, 0))) {
-            mender_log_error("Unable to generate authentication keys");
-            goto END;
+
+        if ((NULL == user_provided_key) || (0 == user_provided_key_length)) {
+            mender_log_error("User provided authentication key is empty");
+            return MENDER_FAIL;
         }
 
-        /* Record keys */
         if (MENDER_OK
-            != (ret = mender_storage_set_authentication_keys(
-                    mender_tls_private_key, mender_tls_private_key_length, mender_tls_public_key, mender_tls_public_key_length))) {
-            mender_log_error("Unable to record authentication keys");
-            goto END;
+            != mender_tls_get_authentication_keys(&mender_tls_private_key,
+                                                  &mender_tls_private_key_length,
+                                                  &mender_tls_public_key,
+                                                  &mender_tls_public_key_length,
+                                                  user_provided_key,
+                                                  user_provided_key_length)) {
+            mender_log_error("Unable to get user provided authentication key");
+            free(user_provided_key);
+            return MENDER_FAIL;
         }
+
+        free(user_provided_key);
+        return MENDER_OK;
+    }
+
+    /* Get keys from store */
+    mender_log_debug("Reading authentication keys from store");
+    switch (mender_storage_get_authentication_keys(
+        &mender_tls_private_key, &mender_tls_private_key_length, &mender_tls_public_key, &mender_tls_public_key_length)) {
+        case MENDER_OK:
+            /* Keys found! */
+            return MENDER_OK;
+        case MENDER_NOT_FOUND:
+            mender_log_debug("Authentication keys not found in store");
+            break;
+        default:
+            mender_log_error("Unable to get authentication keys from store");
+            return MENDER_FAIL;
```

Comment:
Please avoid using `default`. By covering all the cases explicitly, the compiler will warn us if there is a new one not covered. It's easy to list them all above the failure block.

Reply:
I'm not sure if this is what you meant?

Reply:
Yes. Even just this would be good enough for me:
```
case MENDER_DONE:
case MENDER_NOT_IMPLEMENTED:
case MENDER_FAIL:
  mender_log_error(Unable to ...);
  return MENDER_FAIL;
```